sky/services/dynamic/dynamic_service_definition.cc

```diff
@@ -0,0 +1,178 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/services/dynamic/dynamic_service_definition.h"
+
+#include <dlfcn.h>
+
+namespace sky {
+namespace services {
+
+std::unique_ptr<DynamicServiceDefinition> DynamicServiceDefinition::Initialize(
+    const mojo::String& dylib_path) {
+  auto definition = std::unique_ptr<DynamicServiceDefinition>(
+      new DynamicServiceDefinition(dylib_path));
+
+  if (definition->IsReady()) {
+    return definition;
+  }
+
+  return nullptr;
+}
+
+DynamicServiceDefinition::DynamicServiceDefinition(
+    const mojo::String& dylib_path)
+    : service_procs_(), is_ready_(false) {
+  Setup(dylib_path);
+}
+
+DynamicServiceDefinition::~DynamicServiceDefinition() {
+  Teardown();
+}
+
+void DynamicServiceDefinition::Setup(const mojo::String& dylib_path) {
+  std::lock_guard<std::mutex> lock(lifecycle_mtx_);
```

Comment:
what's this lock for? seems pretty scary to be holding a mutex during dylib load/unload depending on what the dylib does. what sort of synchronous calls can the dylib make into the engine with this lock held?